src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -0,0 +1,461 @@
+                        try
+                        {
+                            this.reentrancyDetection.Value = ownedBox = new StrongBox<bool>(true);
+                            await operation().ConfigureAwaitRunInline();
+                        }
+                        finally
+                        {
+                            // Make it clear to any forks of our ExecutionContexxt that the semaphore is no longer owned.
+                            ownedBox.Value = false;
+                            DisposeReleaserNoThrow(releaser);
+                        }
+                    }
+                };
+
+                if (this.joinableTaskFactory != null)
+                {
+                    await this.joinableTaskFactory.RunAsync(semaphoreUser).Task.ConfigureAwaitRunInline();
+                }
+                else
+                {
+                    await semaphoreUser().ConfigureAwaitRunInline();
+                }
+            }
+        }
+
+        /// <summary>
+        /// An implementation of <see cref="ReentrantSemaphore"/> supporting the <see cref="ReentrancyMode.Stack"/> mode.
+        /// </summary>
+        private class StackSemaphore : ReentrantSemaphore
+        {
+            /// <summary>
+            /// The means to recognize that a caller has already entered the semaphore.
+            /// </summary>
+            private readonly AsyncLocal<Stack<StrongBox<AsyncSemaphore.Releaser>>> reentrantCount = new AsyncLocal<Stack<StrongBox<AsyncSemaphore.Releaser>>>();
+
+            /// <summary>
+            /// A flag to indicate this instance was misused and the data it protects should not be touched as it may be corrupted.
+            /// </summary>
+            private bool faulted;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="StackSemaphore"/> class.
+            /// </summary>
+            /// <param name="initialCount">The initial number of concurrent operations to allow.</param>
+            /// <param name="joinableTaskContext">The <see cref="JoinableTaskContext"/> to use to mitigate deadlocks.</param>
+            internal StackSemaphore(int initialCount, JoinableTaskContext joinableTaskContext)
+                : base(initialCount, joinableTaskContext)
+            {
+            }
+
+            /// <inheritdoc />
+            public override async Task ExecuteAsync(Func<Task> operation, CancellationToken cancellationToken = default)
+            {
+                Requires.NotNull(operation, nameof(operation));
+                this.ThrowIfFaulted();
+
+                Stack<StrongBox<AsyncSemaphore.Releaser>> reentrantStack = this.reentrantCount.Value;
+                if (reentrantStack == null || reentrantStack.Count == 0)
+                {
+                    this.reentrantCount.Value = reentrantStack = new Stack<StrongBox<AsyncSemaphore.Releaser>>(capacity: 2);
```

Comment:
>this.reentrantCount.Value = reentrantStack = new Stack<StrongBox<AsyncSemaphore.Releaser>>(capacity: 2); [](start = 20, length = 104)

Is there a race condition here?  Two threads hitting this at the same time will have different instances for ```reentrantStack```.

Reply:
No race condition, because `AsyncLocal<T>` has private storage for each thread and async method (ExecutionContext is forked with copy-on-write semantics).
I've added a code comment to make this clear.